server/src/main/java/org/elasticsearch/search/runtime/IpScriptFieldTermsQuery.java

```diff
@@ -24,16 +28,42 @@ public IpScriptFieldTermsQuery(Script script, IpFieldScript.LeafFactory leafFact
         this.terms = terms;
     }
 
-    @Override
-    protected boolean matches(BytesRef[] values, int count) {
+    boolean matches(BytesRef[] values, int count, BytesRefHash.Finder finder) {
         for (int i = 0; i < count; i++) {
-            if (terms.find(values[i]) >= 0) {
+            if (finder.find(values[i]) >= 0) {
                 return true;
             }
         }
         return false;
     }
 
+    @Override
+    protected boolean matches(BytesRef[] values, int count) {
+        throw new UnsupportedOperationException("This leads to non-thread safe usage of BytesRefHash; use createTwoPhaseIterator instead");
+    }
+
+    boolean matches(IpFieldScript scriptContext, int docId, BytesRefHash.Finder finder) {
+        scriptContext.runForDoc(docId);
+        return matches(scriptContext.values(), scriptContext.count(), finder);
+    }
+
+    protected final TwoPhaseIterator createTwoPhaseIterator(IpFieldScript scriptContext, DocIdSetIterator approximation) {
+        return new TwoPhaseIterator(approximation) {
+            private final BytesRef bytesRef = new BytesRef();
+
+            @Override
+            public boolean matches() {
+                // We need to use a thread safe finder, as this can be called from multiple threads
+                return IpScriptFieldTermsQuery.this.matches(scriptContext, approximation.docID(), terms.newFinder());
```

Comment:
I think we should a single `Finder` per `TwoPhaseIterator`. So instead of passing in `newFinder()` here, we can have a `private final BytesRefHash.Finder finder = terms.newFinder();`

And pass that in to the matches method.